node/src/main/scala/com/wavesplatform/state/Diff.scala

```diff
@@ -145,34 +146,52 @@ case class NewTransactionInfo(transaction: Transaction, affected: Set[Address],
 
-case class LeaseActionInfo(invokeId: ByteStr, dAppPublicKey: PublicKey, recipient: AddressOrAlias, amount: Long)
-
-case class Diff(
-    transactions: collection.Map[ByteStr, NewTransactionInfo] = VectorMap.empty,
-    portfolios: Map[Address, Portfolio] = Map.empty,
-    issuedAssets: Map[IssuedAsset, NewAssetInfo] = Map.empty,
-    updatedAssets: Map[IssuedAsset, Ior[AssetInfo, AssetVolumeInfo]] = Map.empty,
-    aliases: Map[Alias, Address] = Map.empty,
-    orderFills: Map[ByteStr, VolumeAndFee] = Map.empty,
-    leaseState: Map[ByteStr, LeaseDetails] = Map.empty,
-    scripts: Map[Address, Option[AccountScriptInfo]] = Map.empty,
-    assetScripts: Map[IssuedAsset, Option[AssetScriptInfo]] = Map.empty,
-    accountData: Map[Address, AccountDataInfo] = Map.empty,
-    sponsorship: Map[IssuedAsset, Sponsorship] = Map.empty,
-    scriptsRun: Int = 0,
-    scriptsComplexity: Long = 0,
-    scriptResults: Map[ByteStr, InvokeScriptResult] = Map.empty,
-    ethereumTransactionMeta: Map[ByteStr, EthereumTransactionMeta] = Map.empty
+case class Diff private (
+    transactions: Vector[NewTransactionInfo],
+    portfolios: Map[Address, Portfolio],
+    issuedAssets: Map[IssuedAsset, NewAssetInfo],
+    updatedAssets: Map[IssuedAsset, Ior[AssetInfo, AssetVolumeInfo]],
+    aliases: Map[Alias, Address],
+    orderFills: Map[ByteStr, VolumeAndFee],
+    leaseState: Map[ByteStr, LeaseDetails],
+    scripts: Map[Address, Option[AccountScriptInfo]],
+    assetScripts: Map[IssuedAsset, Option[AssetScriptInfo]],
+    accountData: Map[Address, AccountDataInfo],
+    sponsorship: Map[IssuedAsset, Sponsorship],
+    scriptsRun: Int,
+    scriptsComplexity: Long,
+    scriptResults: Map[ByteStr, InvokeScriptResult],
+    ethereumTransactionMeta: Map[ByteStr, EthereumTransactionMeta],
+    transactionFilter: Option[BloomFilter[Array[Byte]]]
 ) {
   @inline
   final def combineE(newer: Diff): Either[ValidationError, Diff] = combineF(newer).leftMap(GenericError(_))
 
+  def transaction(txId: ByteStr): Option[NewTransactionInfo] =
+    if (transactions.nonEmpty && transactionFilter.exists(_.mightContain(txId.arr)))
+      transactions.find(_.transaction.id() == txId)
+    else None
+
+  def withScriptsComplexity(newScriptsComplexity: Long): Diff = copy(scriptsComplexity = newScriptsComplexity)
+
+  def withScriptResults(newScriptResults: Map[ByteStr, InvokeScriptResult]): Diff = copy(scriptResults = newScriptResults)
+
+  def withScriptRuns(newScriptRuns: Int): Diff = copy(scriptsRun = newScriptRuns)
+
   def combineF(newer: Diff): Either[String, Diff] =
     Diff
       .combine(portfolios, newer.portfolios)
-      .map(portfolios =>
+      .map { portfolios =>
+        val newTransactions = if (transactions.isEmpty) newer.transactions else transactions ++ newer.transactions
+        val newFilter = transactionFilter match {
+          case Some(bf) =>
+            newer.transactions.foreach(nti => bf.put(nti.transaction.id().arr))
```

Comment:
Isn't it worth to compare sizes of `this.transactions` and `newer.transactions` to choose bigger bloom filter as the base?

Reply:
We mostly append a diff for a single transaction to an accumulator (e.g. in TransactionDiffer and UtxPoolImpl). Can't think of a place where `newer` has more transactions than `this`.